homeassistant/components/stream/worker.py

```diff
@@ -267,100 +333,55 @@ def stream_worker(
-                packet = next(container_packets)
-                if packet.stream == audio_stream:
-                    found_audio = True
-                elif packet.is_keyframe:  # video_keyframe
-                    first_packet = packet
-                    initial_packets.append(packet)
-            # Get first_dts from subsequent frame to first keyframe
-            while segment_start_dts is None or (
-                audio_stream
-                and not found_audio
-                and len(initial_packets) < PACKETS_TO_WAIT_FOR_AUDIO
-            ):
-                packet = next(container_packets)
-                if packet.stream == audio_stream:
-                    # detect ADTS AAC and disable audio
-                    if audio_stream.codec.name == "aac" and packet.size > 2:
-                        with memoryview(packet) as packet_view:
-                            if packet_view[0] == 0xFF and packet_view[1] & 0xF0 == 0xF0:
-                                _LOGGER.warning(
-                                    "ADTS AAC detected - disabling audio stream"
-                                )
-                                container_packets = filter(
-                                    dts_validator.is_valid,
-                                    container.demux(video_stream),
-                                )
-                                audio_stream = None
-                                continue
-                    found_audio = True
-                elif (
-                    segment_start_dts is None
-                ):  # This is the second video frame to calculate first_dts from
-                    segment_start_dts = packet.dts - packet.duration
-                    first_packet.pts = first_packet.dts = segment_start_dts
-                initial_packets.append(packet)
-            if audio_stream and not found_audio:
-                _LOGGER.warning(
-                    "Audio stream not found"
-                )  # Some streams declare an audio stream and never send any packets
-
-        except (av.AVError, StopIteration) as ex:
-            _LOGGER.error(
-                "Error demuxing stream while finding first packet: %s", str(ex)
+    #
+    # Use a peeking iterator to peek into the start of the stream, ensuring
+    # everything looks good, then go back to the start when muxing below.
+    try:
+        if audio_stream and unsupported_audio(container_packets.peek(), audio_stream):
+            audio_stream = None
+            container_packets = PeekIterator(
+                filter(dts_validator.is_valid, container.demux(video_stream))
             )
-            return False
-        return True
 
-    if not peek_first_dts():
+        # Advance to the first keyframe for muxing, then rewind so the muxing
+        # loop below can consume.
+        first_keyframe = next(filter(is_keyframe, filter(is_video, container_packets)))
+        container_packets.rewind()
```

Comment:
It looks like we only use `rewind()` once, here, to aid with getting us to the first keyframe.
Since we only use it once, it might be clearer (but maybe a little less elegant) to get to the first keyframe down here another way, perhaps by using the peek iterator and then advancing the regular iterator instead

Reply:
I removed rewind, then just had to add a separate call to mux the first keyframe packet.